### DISCUSSION TITLE
Request to support Errors and Exceptions

### DISCUSSION THREAD

**Discussion by FireWizard52** [mikejac/node-red-contrib-google-smarthome#458] [Ideas]
Hi,

As you can make a request, see #280, I would ask, whether it is possible to implement "Error and Exceptions" report
See: https://developers.google.com/assistant/smarthome/reference/errors-exceptions

It  is not a big issue, when you try to switch on a lamp and Google tells you to do so, even if the lamp is already on.
But it would even be better to get voice information "device(s)> <is/are> already on".

it would be a nice improvement, as e.g. Smartnora has such a function implemented already.

Regards

**Comment by Caprico85** [comment]
Hi @FireWizard52,

this would be a very nice addition.

I'm trying to find out how this is done in SmartNora. Is this the async stuff documented in https://github.com/andrei-tatar/node-red-contrib-smartnora/blob/master/doc/nodes/async.md?

Regards

**Comment by FireWizard52** [comment]
Hello @Caprico85,

You asked:

> I'm trying to find out how this is done in SmartNora. Is this the async stuff documented in
> https://github.com/andrei-tatar/node-red-contrib-smartnora/blob/master/doc/nodes/async.md?

I think yes and no.

No because, if I check all the commits in Andrei Tatars repo "nora-firebase-common " the first time async is mentioned is version 1.4.2. (August 23, 2021), while the error code is mentioned for the first time in version 1.0.51 (May 17, 2021)

See: https://github.com/andrei-tatar/nora-firebase-common/commits/master?before=80723bafd1246dffe25b7c2544f5453ca408354a+175&branch=master&qualified_name=refs%2Fheads%2Fmaster

On Google side all information is available at: https://developers.home.google.com/cloud-to-cloud/intents/errors-exceptions and https://developers.home.google.com/cloud-to-cloud/integration/error-handling

And Yes,

Andrei Tatar has introduced asynchronous communication, because, I believe, that makes it possible to use (custom?) response messages.

It is currently experimental, I read, but if you have a Nora Switch, you can tick a box, called Async command execution.
In that case the Nora Switch (or Light) node gets a second output. The second output is used for the error messages.

If that can be a custom message or that it has to be one of the Google List, I'm not sure. I never used it.

The only thing I know, is that, if you use a Nora device and you give a command to switch it "on", you will get a response in case it is already "on", that the device is already "on". I never used the response node. It is long time ago, I used SmartNora in order to test the Google "Local fulfillment" issue. 

![Screenshot_Smart-Nora_Async](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/46283323/b2a6763b-be74-4656-9f15-6db9ee3ff846)

The contents of the upper "Function" node is as follows:

```
//we only replace the payload, to keep the msg id intact

msg.payload = {
    state: {
        on: msg.payload.on,
    },
};

return msg;

```
And the second:

```
//we only replace the payload, to keep the msg id intact

if (msg.payload.on) {
    msg.payload = {
        errorCode: 'alreadyOn',
    };
} else {
    msg.payload = {
        state: {
            on: false,
        },
    };
}

return msg;
```

I hope this info is useful.

Best regards

**Comment by andrei-tatar** [comment]
You can use the async flow to return errors, but there's also a simpler way to do it with NORA:

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/6473183/e3547dec-e220-4fe3-b374-8b386812875c)

**Comment by Caprico85** [comment]
@andrei-tatar The "If state doesn't change, warn user" is a nice idea. Thank you. Should be relatively easy to implement.

Having a "response" node looks to get a custom reponse back into the system is more flexible, but that seems like quite a bit of work.

Don't expect anything from me any time soon. If someone wants to work on a PR, I'd be happy about that too.

**Comment by Caprico85** [comment]
In the meantime I'll convert this into a discussion,
